common/libs/contract/ABIFunction.js

```diff
@@ -0,0 +1,158 @@
+  outputNames;
+  outputTypes;
+  methodSelector;
+  name;
+  payable;
+  type;
+
+  constructor(abiFunc) {
+    Object.assign(this, abiFunc);
+    this._init();
+  }
+
+  _init() {
+    const { inputs, outputs } = this;
+    this.funcParams = this._makeFuncParams();
+    //TODO: do this in O(n)
+    this.inputTypes = inputs.map(({ type }) => type);
+    this.outputTypes = outputs.map(({ type }) => type);
+    this.inputNames = inputs.map(({ name }) => name);
+    this.outputNames = outputs.map(({ name }, i) => name || i);
+
+    this.methodSelector = abi
+      .methodID(this.name, this.inputTypes)
+      .toString('hex');
+  }
+
+  encodeInput = (suppliedInputs = {}) => {
+    const { _processSuppliedArgs, _makeEncodedFuncCall } = this;
+    const args = _processSuppliedArgs(suppliedInputs);
+    const encodedCall = _makeEncodedFuncCall(args);
+    return encodedCall;
+  };
+
+  decodeInput = argString => {
+    const {
+      methodSelector,
+      inputTypes,
+      inputNames,
+      _parsePostDecodedValue
+    } = this;
+
+    // Remove method selector from data, if present
+    argString = argString.replace(`0x${methodSelector}`, '');
+    // Convert argdata to a hex buffer for ethereumjs-abi
+    const argBuffer = new Buffer(argString, 'hex');
+    // Decode!
+    const argArr = abi.rawDecode(inputTypes, argBuffer);
+    //TODO: parse checksummed addresses
+    return argArr.reduce((argObj, currArg, index) => {
+      const currName = inputNames[index];
+      const currType = inputTypes[index];
+      return {
+        ...argObj,
+        [currName]: _parsePostDecodedValue(currType, currArg)
+      };
+    }, {});
+  };
+
+  decodeOutput = argString => {
+    const {
+      methodSelector,
+      outputTypes,
+      outputNames,
+      _parsePostDecodedValue
+    } = this;
+
+    // Remove method selector from data, if present
+    argString = argString.replace(`0x${methodSelector}`, '');
+
+    // Remove 0x prefix
+    argString = argString.replace('0x', '');
+
+    // Convert argdata to a hex buffer for ethereumjs-abi
+    const argBuffer = new Buffer(argString, 'hex');
+    // Decode!
+    const argArr = abi.rawDecode(outputTypes, argBuffer);
+    //TODO: parse checksummed addresses
+    return argArr.reduce((argObj, currArg, index) => {
+      const currName = outputNames[index];
+      const currType = outputTypes[index];
+      return {
+        ...argObj,
+        [currName]: _parsePostDecodedValue(currType, currArg)
+      };
+    }, {});
+  };
+
+  _parsePostDecodedValue = (type, value) => {
+    const { _isBigNumber } = this;
+
+    const valueMapping = {
+      address: val => toChecksumAddress(val.toString(16))
+    };
+
+    return valueMapping[type]
+      ? valueMapping[type](value)
+      : _isBigNumber(value) ? value.toString() : value;
+  };
+
+  _parsePreEncodedValue = (type, value) => {
+    const { _isBigNumber } = this;
+    return _isBigNumber(value) ? value.toString() : value;
+  };
+  _isBigNumber(object) {
```

Comment:
Would move outside of class as a helper utility.

Reply:
agreed